src/hb-utils.c

```diff
@@ -0,0 +1,138 @@
+#include "hb-utils.h"
+
+#include <stdlib.h>
+#include <hb-ot.h>
+
+static hb_variation_t* scan_variation_string(const char* cp1, unsigned int* ret) {
+  hb_variation_t* variations = NULL;
+  hb_variation_t variation;
+  unsigned int nVariations = 0;
+  const char* cp2;
+
+  if (!cp1)
+    return NULL;
+
+  while (*cp1) {
+    if ((*cp1 == ':') || (*cp1 == ';') || (*cp1 == ','))
+      ++cp1;
+    while ((*cp1 == ' ') || (*cp1 == '\t')) /* skip leading whitespace */
+      ++cp1;
+    if (*cp1 == 0)  /* break if end of string */
+      break;
+
+    cp2 = cp1;
+    while (*cp2 && (*cp2 != ':') && (*cp2 != ';') && (*cp2 != ','))
+      ++cp2;
+
+    if (hb_variation_from_string(cp1, cp2 - cp1, &variation)) {
+      variations = realloc(variations, (nVariations + 1) * sizeof(hb_variation_t));
+      variations[nVariations++] = variation;
+    }
+
+    cp1 = cp2;
+  }
+  *ret = nVariations;
+  return variations;
+}
+
+hb_font_t* get_hb_font(lua_State *L, int index) {
+  const char * filename;
+  int face_index = 0;
+  hb_blob_t* blob;
+  hb_face_t* face;
+  hb_font_t* font;
+  unsigned int upem;
+
+  luaL_checktype(L, index, LUA_TTABLE);
+
+  lua_getfield(L, index, "hbFont");
+  if (lua_islightuserdata(L, -1)) { return lua_touserdata(L, -1); }
+
+  lua_getfield(L, index, "filename");
+  if (lua_isstring(L, -1)) { filename = lua_tostring(L, -1); }
+  else { luaL_error(L, "No font filename supplied to get_glyph_dimensions"); }
+
+  lua_getfield(L, index, "index");
+  if (lua_isnumber(L, -1)) { face_index = lua_tointeger(L, -1); }
+
+  blob = hb_blob_create_from_file(filename);
+  face = hb_face_create(blob, face_index);
+  font = hb_font_create(face);
+  upem = hb_face_get_upem(face);
+  hb_font_set_scale(font, upem, upem);
+
+  hb_ot_font_set_funcs(font);
+
+#if HB_VERSION_ATLEAST(3, 3, 0)
```

Comment:
If we want to apply variations only when the subsetter is present, we can do:
```suggestion
#ifdef HAVE_HARFBUZZ_SUBSET
```

Reply:
I have done a slightly different variation of this on 2f71ef4d90b14dc234d6c6c6efbc55ee3eac7332, but I’m not sure it is making any difference with the failed test.

Reply:
Thanks. I think this is the direction I'd lake to take it. From a troubleshooting and support standpoint I think it will be clearer what to expect if we deliberately turn this on or off rather than just letting it try. Builders are going to start getting errors for HB<6 now anyway and need to explicitly disable features to continue. They will be told some things won't work at that point, and then I want to add an error when rendering documents that request access to those features. I think lumping all variations support behind the same feature flag is reasonable. I might reword the flag and help messages with that in mind though.

Reply:
I have done this part now.